src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -14,18 +14,66 @@ namespace System.Text.Json.Serialization.Metadata
     /// <summary>
     /// Provides JSON serialization-related metadata about a type.
     /// </summary>
-    /// <remarks>This API is for use by the output of the System.Text.Json source generator and should not be called directly.</remarks>
     [DebuggerDisplay("{DebuggerDisplay,nq}")]
-    [EditorBrowsable(EditorBrowsableState.Never)]
-    public partial class JsonTypeInfo
+    public abstract partial class JsonTypeInfo
     {
         internal const string JsonObjectTypeName = "System.Text.Json.Nodes.JsonObject";
 
-        internal delegate object? ConstructorDelegate();
-
         internal delegate T ParameterizedConstructorDelegate<T, TArg0, TArg1, TArg2, TArg3>(TArg0 arg0, TArg1 arg1, TArg2 arg2, TArg3 arg3);
 
-        internal ConstructorDelegate? CreateObject { get; set; }
+        private JsonPropertyDictionary<JsonPropertyInfo>.ValueList? _properties;
+
+        /// <summary>
+        /// Object constructor. If set to null type is not deserializable.
+        /// </summary>
+        public Func<object>? CreateObject
+        {
+            get => _createObject;
+            set
+            {
+                SetCreateObject(value);
+            }
+        }
+
+        private protected abstract void SetCreateObject(Delegate? createObject, bool useForExtensionDataProperty = false);
+        private protected Func<object>? _createObject;
+
+        // this is only assigned if Kind == None
+        internal Func<object>? CreateObjectForExtensionDataProperty { get; set; }
+
+        /// <summary>
+        /// Gets JsonPropertyInfo list. Only applicable when Kind is Object.
+        /// </summary>
+        public IList<JsonPropertyInfo> Properties
+        {
+            get
+            {
+                if (_properties != null)
+                {
+                    return _properties;
+                }
+
+                if (Kind == JsonTypeInfoKind.Object)
```

Comment:
Is the getter called in the hot path at all? Wondering if it might be worth factoring all the initialization code into a separate method to make the getter inlineable.

Reply:
No, we only use PropertyCache in the hot path, this is only for users to see, IMO not worth to optimize